### ISSUE TITLE
order of args for merge

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/phet-info#121]
Adding to the discussion of the `options/config` design pattern in #88 

I encountered this anti-pattern in https://github.com/phetsims/projectile-motion/issues/197#issuecomment-549055000, and it's not the first time that I've seen it:

```js
class MySubclass extends SomeClass {
  constructor( ..., options ) {
    options = merge( {
      // values specific to MySubclass  
    }, SomeConstants.SOME_CLASS_OPTIONS, options );
    ...
  }
}
```

The problem here is the order of args to `merge`. It makes it possible for a change to `SomeConstants.SOME_CLASS_OPTIONS` to unintentionally change the behavior of `MySubclass`.  

To avoid this problem, args should be ordered from least to most specific. For example, the above should be order like this:  (Note the first arg is `{}` to prevent modification of `SOME_CLASS_OPTIONS`.)
```js
class MySubclass extends SomeClass {
  constructor( ..., options ) {
    options = merge( {}, SomeConstants.SOME_CLASS_OPTIONS, {
      // values specific to MySubclass  
    }, options );
    ...
  }
}
```

**Comment by pixelzoom** [comment]
@Denz1994 would you please add this to the list of issues to discuss at 11/4/19 design patterns meeting?

**Comment by pixelzoom** [comment]
Another example can be found in `WAVE_INTERFERENCE/DiffractionNumberControl`:

```js
      const mergedOptions = merge( {
        sliderOptions: {
          majorTicks: [ {
            value: property.range.min,
            label: new WaveInterferenceText( property.range.min.toFixed( 2 ) )
          }, {
            value: property.range.max,
            label: new WaveInterferenceText( property.range.max.toFixed( 2 ) )
          } ]
        }
      }, options, WaveInterferenceConstants.NUMBER_CONTROL_OPTIONS );
```

If I were to add a default value for `sliderOptions.majorTicks` to `WaveInterferenceConstants.NUMBER_CONTROL_OPTIONS`, that would change the behavior of `DiffractionNumberControl`.

**Comment by pixelzoom** [comment]
Shall we add this to the [Best Practices ](https://github.com/phetsims/phet-info/blob/master/doc/phet-software-design-patterns.md#best-practices) section of options/config pattern? 

> (4) When overriding options/config using `merge`, arguments should be ordered from least to most specific.  For example:

```js
// Incorrect
class MyPanel extends Panel {
  constructor( options ) {
    options = merge( options, {
      fill: 'yellow'
      ...
    }, MyConstants.PANEL_OPTIONS );
    ...
  }
}

// Correct
class MyPanel extends Panel { 
  constructor( options ) {
    options = merge( {}, MyConstants.PANEL_OPTIONS, {
      fill: 'yellow'
      ...
    }, options);
    ...
  }
}
```

**Comment by Denz1994** [comment]
As part of the design patterns discussion on 11/04/19:

The`required`  method described in https://github.com/phetsims/phet-info/issues/115 should be used to validate the required fields in the config object. The validation must be provided **after** the config object is merged to confirm that it isn't overridden by the other merge args. See the below example here:
```
class MyPanel extends Panel { 
  constructor( config ) {
    config = merge( { }, MyConstants.PANEL_DEFAULTS, {
      fill: required( config.fill )
    }, config);
  }
}
```

This ensures the require() validation targets all necessary fields from the passed in the config object.

**Comment by Denz1994** [comment]
Devs have been made aware of this issue at our patterns-design meeting. There is nothing left for this issue, so I'm closing.

**Comment by pixelzoom** [comment]
Reopening. I'd like to add this to "Best Practices" for options/config in the patterns doc.

**Comment by pixelzoom** [comment]
Over in https://github.com/phetsims/phet-info/issues/117#issuecomment-549906410:

> > (4)Best Practices Point 4: Is this handled by PhET_Core/merge.js validateMergableObject() assertion?
> Yes.

So I deleted best practice (4).

Relevant to this issue, I added new best practice (4) as proposed in https://github.com/phetsims/phet-info/issues/121#issuecomment-549435679.

I also noticed that best practice (3) was a bit of a mess, and contained some outright errors. So I revised it.

Back to @Denz1994 for review.
